### PR TITLE
Add reset support for IIS

### PR DESCRIPTION
**Pull request by jkotalik**
Part of support gRPC in IIS.

This currently is targeting the Trailers PR for now.

@JamesNK this may be relevant, but from what I can tell, if you call WriteAsync and then call Reset(), it isn't guaranteed that the data written in WriteAsync is sent to the client. Is this a large concern?

### CODE REVIEW

**Comment by JamesNK** [comment]
> @JamesNK this may be relevant, but from what I can tell, if you call WriteAsync and then call Reset(), it isn't guaranteed that the data written in WriteAsync is sent to the client. Is this a large concern?

No. I can't think of a reason why it would be a problem for gRPC.

Something to watch out for is multi-threading issues. CompleteAsync + Reset could be called from another thread at the same time a WriteAsync is happening. Reset waiting until the write completing is fine, just want to make sure it doesn't explode 😄

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs [hunk 1]
```diff
@@ -394,6 +395,28 @@ IHeaderDictionary IHttpResponseTrailersFeature.Trailers
...
+            {
+                return this;
+            }
+
+            return null;
+        }
+
+        void IHttpResetFeature.Reset(int errorCode)
+        {
+            SetResetCode(errorCode);
```
Comment (JamesNK) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs:411]: Validate that errorCode isn't negative? I don't know if we did that in Kestrel or HttpSys

File: src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs [hunk 2]
```diff
@@ -1135,6 +1135,168 @@ public Task ResponseTrailers_NullValues_Ignored(HttpContext context)
...
+            {
+                Assert.Equal("HTTP/2", httpContext.Request.Protocol);
+                var feature = httpContext.Features.Get<IHttpResetFeature>();
+                Assert.NotNull(feature);
+
+                var read = await httpContext.Request.Body.ReadAsync(new byte[10], 0, 10);
+                Assert.Equal(10, read);
+
+                var readTask = httpContext.Request.Body.ReadAsync(new byte[10], 0, 10);
+                feature.Reset(1111);
```
Comment (JamesNK) [inline src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs:1284]: Add a test that `Reset(0)` works. Initial HttpSys implementation incorrectly didn't allow it. I want to double check it works

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs (hunk 1 above)
Comment (JamesNK) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs:411]: https://http2.github.io/http2-spec/#ErrorCodes

Size can be up to 32 bit so there is no need to check for an upper limit.

Comment (jkotalik) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs:411]: I don't this HttpSys does, but I can add that check.

File: src/Servers/IIS/IIS/src/Core/IISHttpServer.cs [hunk 3]
```diff
@@ -202,9 +202,9 @@ private static void OnRequestsDrained(IntPtr serverContext)
             {
                 server = (IISHttpServer)GCHandle.FromIntPtr(serverContext).Target;
 
-                server._nativeApplication.StopCallsIntoManaged();
-                server._shutdownSignal.TrySetResult(null);
-                server._cancellationTokenRegistration.Dispose();
+                server?._nativeApplication.StopCallsIntoManaged();
```
Comment (davidfowl) [inline src/Servers/IIS/IIS/src/Core/IISHttpServer.cs:205]: This can be be null?

Comment (jkotalik) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs:205]: yeah noticed when debugging after the process is shutting down when I was attached to the debugger.

File: src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs [hunk 4]
```diff
@@ -1135,6 +1135,191 @@ public Task ResponseTrailers_NullValues_Ignored(HttpContext context)
...
+                _Reset_BeforeRequestBody_ResetsCt.SetException(ex);
+            }
+        }
+
+        public async Task Reset_BeforeRequestBody_Resets_Complete(HttpContext httpContext)
+        {
+            await _Reset_BeforeRequestBody_ResetsCt.Task;
+        }
+
+        private TaskCompletionSource<object> _Reset_DuringRequestBody_ResetsCt = new TaskCompletionSource<object>();
```
Comment (davidfowl) [inline src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs:1293]: why is the casing here crazy?

File: src/Servers/IIS/IIS/src/Core/IISHttpServer.cs (hunk 3 above)
Comment (halter73) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs:205]: It looks like we complete the Task returned by StopAsync() as soon as CancellationToken fires. This means Hosting can dispose the IISHttpServer (which frees the GCHandle) before OnRequestsDrained gets called.

I bet it's not just OnRequestsDrained callback where the server reference can wind up being null. We also should be null checking the server reference in HandleRequest and HandleShutdown. I'm not sure if there's anywhere else this is necessary.

Comment (davidfowl) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs:205]: This makes me uneasy. I'd like to better understand this before we start checking for null everywhere. Should we fix the draining logic?

File: src/Servers/IIS/IIS/src/Core/IISHttpServer.cs [hunk 5]
```diff
@@ -135,7 +135,7 @@ private static NativeMethods.REQUEST_NOTIFICATION_STATUS HandleRequest(IntPtr pI
                 // Unwrap the server so we can create an http context and process the request
                 server = (IISHttpServer)GCHandle.FromIntPtr(pvRequestContext).Target;
 
-                var context = server._iisContextFactory.CreateHttpContext(pInProcessHandler);
+                var context = server?._iisContextFactory.CreateHttpContext(pInProcessHandler);
```
Comment (davidfowl) [inline src/Servers/IIS/IIS/src/Core/IISHttpServer.cs:138]: This null check looks bogus. You'll get a null context back then an exception will be thrown when you QUWI(null)

File: src/Servers/IIS/IIS/src/Core/IISHttpServer.cs (hunk 3 above)
Comment (jkotalik) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs:205]: I feel uneasy with this too. Potentially we should setup a meeting to go over this logic?

Comment (davidfowl) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs:205]: yah

**Comment by Tratcher** [comment]
> > @JamesNK this may be relevant, but from what I can tell, if you call WriteAsync and then call Reset(), it isn't guaranteed that the data written in WriteAsync is sent to the client. Is this a large concern?
> 
> No. I can't think of a reason why it would be a problem for gRPC.
> 
> Something to watch out for is multi-threading issues. CompleteAsync + Reset could be called from another thread at the same time a WriteAsync is happening. I don't care what happens to the write - we can wait until it is finished or interrupt it. I just want to make sure everything doesn't explode 😄

@JamesNK that breaks our guidance around HttpContext thread safety (or lack thereof). If that's something gRPC is doing then it needs to add its own locking.

File: src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs [hunk 6]
```diff
@@ -1135,6 +1135,191 @@ public Task ResponseTrailers_NullValues_Ignored(HttpContext context)
...
+        {
+            await httpContext.Response.Body.FlushAsync();
+            throw new Exception("Application exception");
+        }
+
+        public async Task AppException_AfterHeaders_ResetInternalError(HttpContext httpContext)
+        {
+            await httpContext.Response.Body.FlushAsync();
+            throw new Exception("Application exception");
+        }
```
Comment (Tratcher) [inline src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs:1153]: Can you combine some of these? The app code is the same, only the behavior observed on the client is different.

**Review by Tratcher — APPROVED** [review]

File: src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs (hunk 6 above)
Comment (jkotalik) [reply src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs:1153]: I purposely didn't combine them to keep them 1 to 1 with the tests. IMO it makes it easier to match test to the part of the startup.

**Pull request merged**